### PR TITLE
Boxes resolving fix

### LINKED ISSUE #567: Wallet: resolve box
Looks like box resolving is incorrect in Wallet - it resolve boxes, but failes to spend it in transaction generator. 

```
09:58:05.187 DEBUG [ctor.default-dispatcher-3] o.e.n.w.ErgoWalletActor - Uncertain box is mine! TrackedBox {"creationHeight":3230,"certain":false,"creationTransactionId":"8a3bd1276aab49061e6
9a172c9221e94d9e6b14eb6cb3ebe63da094874339230","spendingTransactionId":null,"onchain":true,"creationOutIndex":1,"box":{"boxId":"dade75858d7fcef86761df81f7ec63ee51ddc065e6b4227cea1f0c94802ab3
42","value":75000000000,"proposition":"9683020192a39a8cc7a70105a00bcd07034f03bd8e1e84e925aa3a8eef95dc0284ff5625419b01db30392afdd4a9f784c1","assets":[],"creationHeight":3230,"additionalRegist
ers":{}},"spendingHeight":null,"spent":false}
09:58:05.187 DEBUG [ctor.default-dispatcher-3] o.e.n.wallet.WalletStorage - New Onchain Certain box arrived: TrackedBox {"creationHeight":3230,"certain":true,"creationTransactionId":"8a3bd12
76aab49061e69a172c9221e94d9e6b14eb6cb3ebe63da094874339230","spendingTransactionId":null,"onchain":true,"creationOutIndex":1,"box":{"boxId":"dade75858d7fcef86761df81f7ec63ee51ddc065e6b4227cea
1f0c94802ab342","value":75000000000,"proposition":"9683020192a39a8cc7a70105a00bcd07034f03bd8e1e84e925aa3a8eef95dc0284ff5625419b01db30392afdd4a9f784c1","assets":[],"creationHeight":3230,"addi
tionalRegisters":{}},"spendingHeight":null,"spent":false}
09:58:05.187 DEBUG [ctor.default-dispatcher-3] o.e.n.wallet.WalletStorage - Update balance with UNSPENT Onchain dade75858d7fcef86761df81f7ec63ee51ddc065e6b4227cea1f0c94802ab342 TrackedBox {"
creationHeight":3230,"certain":true,"creationTransactionId":"8a3bd1276aab49061e69a172c9221e94d9e6b14eb6cb3ebe63da094874339230","spendingTransactionId":null,"onchain":true,"creationOutIndex":
1,"box":{"boxId":"dade75858d7fcef86761df81f7ec63ee51ddc065e6b4227cea1f0c94802ab342","value":75000000000,"proposition":"9683020192a39a8cc7a70105a00bcd07034f03bd8e1e84e925aa3a8eef95dc0284ff562
5419b01db30392afdd4a9f784c1","assets":[],"creationHeight":3230,"additionalRegisters":{}},"spendingHeight":null,"spent":false} balance: 8025000000000 Map(), total: 8100000000000 Map()
09:58:05.187 DEBUG [ctor.default-dispatcher-3] o.e.n.w.ErgoWalletActor - Failed to resolve uncertainty for dae0a8015bc79fff2b25d6dac757c356f7dc4e28ffecc0fa135fa7c70111e1d3 created at Some(40
41) while current height is 4232
09:58:05.196 DEBUG [r.default-dispatcher-1157] o.e.local.ErgoMiner - Got candidate block at height 4233 with 1 transactions
09:58:05.211 DEBUG [r.default-dispatcher-1323] o.e.n.UtxoNodeViewHolder - Cache size after: 512
09:58:05.211 DEBUG [r.default-dispatcher-1323] o.e.n.UtxoNodeViewHolder - Cache size before: 513
09:58:05.235 DEBUG [r.default-dispatcher-1323] o.e.n.UtxoNodeViewHolder - Cache size after: 512
09:58:05.235 DEBUG [r.default-dispatcher-1157] o.e.n.ErgoDeliveryTracker - Set modifier 7b4e69dbc88c81c199ee6bce246ab1a6ae62269533c9daf622275b7cf9ca99ae from status Received to status Unknow
n.
09:58:05.236 INFO  [r.default-dispatcher-1323] o.e.n.UtxoNodeViewHolder - Got locally generated modifier d8bc0ac22f546c3ab1dcc9d44665e40e1b3eadb9ba44026b8261187a11e2d974 of type 104
09:58:05.236 INFO  [r.default-dispatcher-1323] o.e.n.UtxoNodeViewHolder - Apply modifier d8bc0ac22f546c3ab1dcc9d44665e40e1b3eadb9ba44026b8261187a11e2d974 of type 104 to nodeViewHolder
09:58:05.236 DEBUG [r.default-dispatcher-1323] o.e.n.h.ErgoHistory$$anon$2 - Trying to append modifier d8bc0ac22f546c3ab1dcc9d44665e40e1b3eadb9ba44026b8261187a11e2d974 of type 104 to history
09:58:05.236 DEBUG [r.default-dispatcher-1323] o.e.n.UtxoNodeViewHolder - Going to apply modifications to the state: ProgressInfo(BranchPoint: None,  to remove: List(), to apply: List())
09:58:05.236 DEBUG [r.default-dispatcher-1157] o.e.n.ErgoDeliveryTracker - Set modifier d8bc0ac22f546c3ab1dcc9d44665e40e1b3eadb9ba44026b8261187a11e2d974 from status Unknown to status Held.
09:58:05.237 DEBUG [r.default-dispatcher-1323] o.e.l.TransactionGenerator - Failed to generate tx: No enough boxes to assemble a transaction for List(ErgoBoxCandidate(55800, ProveDlog(Consta
ntNode((397e61d23487aa516955d9175ff3b6c69a7d72f6aac3de8383577fd107d9e4a9,5df0133aa60ec519d9c7016a4ac579a43ed45eadc92b7ab1160aaaeba287c6ba,16bbb2383489815ebd4361c03f63943012165890e59d29b99df3
2d300807d2e2,306072d820c3c957c8648ec9d88507213a3d963383e0b939ea6c0043ec2a00),SGroupElement)),tokens: (00b2a6bd344c29298fd9d01c7691aef8c6ca3076ec83f9e6adc98fd43ec17405:400000000), Map(org.erg
oplatform.ErgoBox$R4$@6a34cb0c -> ConstantNode(52513E08,SString), org.ergoplatform.ErgoBox$R5$@1bb3e6c2 -> ConstantNode(52513E08 description,SString), org.ergoplatform.ErgoBox$R6$@757a243d -
> ConstantNode(10,SInt)), creationHeight: 4232), ErgoBoxCandidate(100000, AND(ConcreteCollection(Vector(EQ(Height,SelectField(ExtractCreationInfo(ByIndex(Outputs,ConstantNode(0,SInt),None)),
1)), EQ(ExtractScriptBytes(ByIndex(Outputs,ConstantNode(0,SInt),None)),Append(ConstantNode([B@567c7b04,Array[SByte]),MinerPubkey)), EQ(SizeOf(Outputs),ConstantNode(1,SInt))),SBoolean)),token
s: (), Map(), creationHeight: 4232))
```

### PR DESCRIPTION
**Pull request by catena2w**
Close #567

### CODE REVIEW

File: src/main/scala/org/ergoplatform/nodeView/wallet/WalletStorage.scala [hunk 1]
```diff
@@ -51,14 +51,14 @@ class WalletStorage extends ScorexLogging {
 
   private def put(trackedBox: TrackedBox): Option[TrackedBox] = {
     if (trackedBox.certainty == Uncertain) uncertainBoxes += trackedBox.boxId
-    if (trackedBox.spendingStatus == Unspent) unspentBoxes += trackedBox.boxId
+    if (trackedBox.spendingStatus == Unspent && trackedBox.certainty == Certain) unspentBoxes += trackedBox.boxId
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/nodeView/wallet/WalletStorage.scala]:
```
if (trackedBox.certainty == Uncertain) uncertainBoxes += trackedBox.boxId
else if (trackedBox.spendingStatus == Unspent) unspentBoxes += trackedBox.boxId
```
will look better

File: src/main/scala/org/ergoplatform/nodeView/wallet/WalletStorage.scala [hunk 2]
```diff
@@ -51,14 +51,14 @@ class WalletStorage extends ScorexLogging {
...
     registry.put(trackedBox.boxId, trackedBox)
   }
 
   private def remove(boxId: ModifierId): Option[TrackedBox] = {
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/nodeView/wallet/WalletStorage.scala]: Looks like all we need here is 
```
    uncertainBoxes.remove(boxId)
    unspentBoxes.remove(boxId)
    registry.remove(boxId)
```

Comment (oskin1) [reply src/main/scala/org/ergoplatform/nodeView/wallet/WalletStorage.scala]: @catena2w, why making potentially redundant function calls?

File: src/main/scala/org/ergoplatform/nodeView/wallet/WalletStorage.scala (hunk 1 above)
Comment (oskin1) [reply src/main/scala/org/ergoplatform/nodeView/wallet/WalletStorage.scala]: @catena2w, done

File: src/main/scala/org/ergoplatform/local/TransactionGenerator.scala [hunk 3]
```diff
@@ -88,13 +88,13 @@ class TransactionGenerator(viewHolder: ActorRef,
     val payloadReq: Future[Option[TransactionRequest]] = wallet.confirmedBalances().map { balances =>
       Random.nextInt(100) match {
         case i if i < 70 =>
-          Some(PaymentRequest(randProposition, randAmount, None, None))
+          Some(PaymentRequest(randProposition, math.min(randAmount, balances.balance), None, None))
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/local/TransactionGenerator.scala]: I think `balances.balance - feeReq.value` will be correct here

File: src/test/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSpec.scala [hunk 4]
```diff
@@ -617,6 +619,41 @@ class ErgoWalletSpec extends PropSpec with WalletTestOps {
...
+
+      wallet.scanPersistent(modifiedBlock)
+
+      blocking(Thread.sleep(100))
+
+      val requestWithTotalAmount = PaymentRequest(
+        ErgoAddressEncoder(0: Byte).fromProposition(pubKey).get, totalAvailableAmount, None, None)
+      val requestWithCertainAmount = requestWithTotalAmount.copy(value = certainAmount)
+
+      Async.async {
```
Comment (catena2w) [inline src/test/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSpec.scala]: You may just write
```
      await(wallet.generateTransaction(Seq(requestWithTotalAmount))) shouldBe 'failure
      await(wallet.generateTransaction(Seq(requestWithCertainAmount))) shouldBe 'success
```

File: src/test/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSpec.scala [hunk 5]
```diff
@@ -617,6 +619,41 @@ class ErgoWalletSpec extends PropSpec with WalletTestOps {
     }
   }
 
+  property("only certain and unspent boxes is used for transaction generation") {
```
Comment (catena2w) [inline src/test/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSpec.scala]: This test pass with old wallet storage. It should fail with old WalletStorage, and pass with the new one

Comment (oskin1) [reply src/test/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSpec.scala]: @catena2w, already fixed.

File: src/test/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSpec.scala (hunk 4 above)
Comment (oskin1) [reply src/test/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSpec.scala]: @catena2w, fixed

File: src/main/scala/org/ergoplatform/local/TransactionGenerator.scala (hunk 3 above)
Comment (oskin1) [reply src/main/scala/org/ergoplatform/local/TransactionGenerator.scala]: @catena2w, fixed.

**Pull request merged**